### ISSUE TITLE
Using two tabs at the same time as different people gets the system tied in knots

### ISSUE THREAD

**Issue by akc42** [akc42/football-mobile#72]
I guess this must be related to cookies which are in essence site based rather than perhaps a token which could be tab based.

So plan to solve this is
- [x] consider what criteria I need related to remembering who I am between accesses and tabs
- [x] based on the criteria I come up with - look again at Jake and Sumar's discussion about communication between tabs - figure out how to implement
- [ ] implement

**Comment by akc42** [comment]
Thinking about the criteria

- When running in a tab _ should take on the identity that I signed on with along with all the criteria that entails.
- I need to know who I am signed in as - somewhere in the header maybe. - Although better might be the tabs (is this cross browser)
- When I totally refresh the page (tab) I should remember who I am if I have not logged out
- closing the tab should loose that identity unless I selected remember me at log in (AND THERE IS NO OTHER TAB OPEN)

**Comment by akc42** [comment]
One possible solution to refresh is to record time of close in local storage, then read it on reload.  Use some random smallish number like 10 seconds to determine it was a refresh and not a reload.

**Comment by akc42** [comment]
It is now looking like session storage maybe the way to go.  It survives reloads.  I might be able to use it for `global.user` and is cleared when a browser tab is closed, but not when it is reloaded.

We can also load data in local storage - so if we then check that the same user is not already open in another window at some point. (we can use weblockApi to avoid race conditions in this process).  If its the same user, we could? synchronise their log in/log out.

Now we need the server to know we are authorised to access certain apis.  We could use cookies for that - or we could use "/api/:token/... for that  .  The issue with cookies is that they get sent with every request, where as we only really need them sent for the api requests.  (and not even all of them - some such as config are available to anyone).

Change a tabs title by writing to `document.title`. "Football (user)"

**Comment by akc42** [comment]
so plan of action is
- use localstorage to hold value of "remembered" user as a jwt with a long expiry date.  This will be returned the first time when the user successfully logs in and has the remember me set on the password form
- use localstorage to hold the number of tabs open
- use web lock to update local storage on loading a tab and on unloading a tab.
- only use the remembered user if this is the first tab open.
- use session storage to hold `global.user`
-write logged on user to title with document.title = 
- do away with cookies.  Api that need authorisation calls will have a token addeded

**Comment by akc42** [comment]
Problem - web locks isn't obviously supported on ios etc.  need to find a way to avoid that.

I think the way round this is using service - worker and the broadcast mechanism to send a query to the service worker as to whether I am the first client or not.  If I am it can tell me in the reply.

**Comment by akc42** [comment]
It turns out service worker is not necessary and is to wait for another issue.  We simply used the storage Event to have the open tabs send a reply to the I am a new tab opening.  If that new tab hears a reply within a timeout (70ms) it knows it is not alone and therefore is not the first tab - and forces a new login.

We did all this inside global so the majority of it is hidden from the rest of the code.  the api function appends the token (stored in sessionStorage (copied from local storage) to every api call.  Its no bigger than a cookie and doesn't need to be sent for any static files.
